### ISSUE TITLE
High Gamma Dataset Deep4Net results reproduction

### ISSUE THREAD

**Issue by martinwimpff** [braindecode/braindecode#395]
I cannot reproduce the results of the Deep4Net with the HGD.

The accuracy should be at around 90% (for subject 1). I am able to reproduce these results with the ShallowNet but not with the DeepNet.
Is there anything special to the training, optimization, regularization, etc. of the DeepNet compared to the ShallowNet?
@robintibor: any ideas?

I used the same hyperparameters as in the 2017 publication. The only thing I changed in both cases (ShallowNet and DeepNet) is that I used Conv2dWithContraint instead of Conv2d to add extra regularization:
```
class Conv2dWithConstraint(nn.Conv2d):
    def __init__(self, *args, max_norm=1, dim=0, **kwargs):
        self.max_norm = max_norm
        self.dim = dim
        super(Conv2dWithConstraint, self).__init__(*args, **kwargs)

    def forward(self, x):
        self.weight.data = torch.renorm(
            self.weight.data, p=2, dim=self.dim, maxnorm=self.max_norm
        )
        return super(Conv2dWithConstraint, self).forward(x)
```

The rest of the code below borrows from the [BCICIV2a_example in this repository](https://github.com/braindecode/braindecode/blob/master/examples/plot_bcic_iv_2a_moabb_trial.py) as well as from the [original HGD repo](https://github.com/robintibor/high-gamma-dataset/tree/ff3932852a9aa0e532b9fc9bdcea2934d2e33e55).

Code is available at: https://colab.research.google.com/drive/1tpStIcOFyZNa8Z2ePBxxJztqzvC0qcNl?usp=sharing

**Comment by robintibor** [comment]
Hi, thanks for your interest!
so there is code to reproduce the HGD results here: https://gist.github.com/robintibor/6a95a85088e651392c1bb4f912d1528e#file-run_and_train_amp_grad-py-L242-L256
As you may notice, this is using cropped training. For trialwise/noncropped training, the current pipeline will get worse results than the original paper (original superold code at https://github.com/robintibor/braindevel/), as we have not optimized the hyperparameters for the changed pipeline for the trialwise case.

**Comment by martinwimpff** [comment]
Hi @robintibor, thanks for your answer!
Do you still have the original tuned HP of the paper? 
Or can you guide me through the old code so that I roughly know where to look ;)

**Comment by robintibor** [comment]
Hm so hyperparams might be scattered among those files:
https://github.com/robintibor/braindevel/blob/21f58aa74fdd2a3b03830c950b7ab14d44979045/braindecode/configs/experiments/paper/bci_competition/cnt/deep_4_cnt_net_car.yaml / https://github.com/robintibor/braindevel/blob/21f58aa74fdd2a3b03830c950b7ab14d44979045/braindecode/configs/experiments/4sec_movements/cnt/cnt_4_layers_simple.yaml
https://github.com/robintibor/braindevel/blob/21f58aa74fdd2a3b03830c950b7ab14d44979045/braindecode/configs/experiments/4sec_movements/cnt/defaults.yaml
Since I cannot see them there, assume we used default learning rate (1e-3) and weight decay 0 for Adam.

In general though, I would recommend to rather try tuning the hyperparams with the new pipeline (training on all data and using cosine annealing), I would expect the overall result to be better.

**Comment by martinwimpff** [comment]
Thanks @robintibor! I was able to reproduce and partly surpass the results of the paper (~74% overall test acc).
The configuration that worked best for me was this one:

- general/preprocessing
  - n_epochs: 1000
  - sfreq: 250Hz
  - BP: 0/0.5-38Hz
  - slice: 0.5s before start - 0.5s before end (4s)
  - batch size 64
  - lr: 1e-3
  - scheduler: cosine annealing
- model
  - filter_length: 10
  - pool length/stride: 3
  - dropout: 50%
  - batch_norm
  - elu 
  - max_norm: 2.0 and 0.5 (classifier)

**Comment by bruAristimunha** [comment]
Hi again @martinwimpff!

I see that you opened and closed the issue. It looks like you solve the reproduction point with robin's instruction. Am I right? Can we close?

**Comment by martinwimpff** [comment]
Hi @bruAristimunha, I closed it by accident ;)
I ran some experiments today and got at least in the right range (86-87% overall accuracy). The gap to ShallowNet remains quite large (overall accuracy 93%). I think one would have to run some more experiments to get a really good configuration.
But as we are in roughly the same range, I am okay with closing this issue.

If anyone finds a configuration, that matches the published results, feel free to share it here ;)
